crates/sui-framework/packages/sui-framework/sources/crypto/zklogin.move

```diff
@@ -0,0 +1,111 @@
+
+        /// The issuer
+        iss: String,
+
+        /// The audience (wallet)
+        aud: String,
+    }
+
+    /// Verify that the caller's address was created using zklogin and the given parameters. If so,
+    /// a VerifiedID cantaining the caller's id (claim name and value, issuer and wallet id) is
+    /// returned. Otherwise, None is returned.
+    public fun verify_zklogin_id(
+        ctx: &TxContext,
+        kc_name: String,
+        kc_value: String,
+        iss: String,
+        aud: String,
+        pin_hash: u256
+    ): Option<VerifiedID> {
+        if (check_zklogin_id(tx_context::sender(ctx), &kc_name, &kc_value, &iss, &aud, pin_hash)) {
+            option::some(VerifiedID {kc_name, kc_value, iss, aud})
+        } else {
+            option::none()
+        }
+    }
+
+    /// Returns true if `address` was created using zklogin and the given parameters.
+    public fun check_zklogin_id(
+        address: address,
+        kc_name: &String,
+        kc_value: &String,
+        iss: &String,
+        aud: &String,
+        pin_hash: u256
+    ): bool {
+        check_zklogin_id_internal(address,
+            std::string::bytes(kc_name),
+            std::string::bytes(kc_value),
+            std::string::bytes(iss),
+            std::string::bytes(aud),
+            pin_hash)
+    }
+
+    /// Returns true if `address` was created using zklogin and the given parameters.
+    native fun check_zklogin_id_internal(
+        address: address,
+        name: &vector<u8>,
+        value: &vector<u8>,
+        iss: &vector<u8>,
+        aud: &vector<u8>,
+        pin_hash: u256
+    ): bool;
+
+    /// Posession of a VerifiedIssuer proves that the uses's address was created using zklogin and
+    /// with the given issuer (identity provider).
+    struct VerifiedIssuer {
+        /// The issuer
+        iss: String,
+    }
+
+    /// Verify that the caller's address was created using zklogin with the given issuer. If so,
+    /// a VerifiedIssuer object cantaining the issuers id returned. Otherwise, None is returned.
+    public fun verify_zklogin_iss(
+        ctx: &mut TxContext,
+        address_seed: u256,
+        iss: String,
+    ): Option<VerifiedIssuer> {
+        if (check_zklogin_iss(tx_context::sender(ctx), address_seed, &iss)) {
+            option::some(VerifiedIssuer {iss})
+        } else {
+            option::none()
+        }
+    }
+
+    /// Returns true if `address` was created using zklogin with the given issuer and address seed.
+    public fun check_zklogin_iss(
+        address: address,
+        address_seed: u256,
+        iss: &String,
+    ): bool {
+        check_zklogin_iss_internal(address, address_seed, std::string::bytes(iss))
+    }
+
+    /// Returns true if `address` was created using zklogin with the given issuer and address seed.
+    native fun check_zklogin_iss_internal(
+        address: address,
+        address_seed: u256,
+        iss: &vector<u8>,
+    ): bool;
+}
```

Comment:
- We should add getters for each of the fields in `VerifiedID` and `VerifiedIssuer`
- I think functions to that you delete `VerifiedID` and `VerifiedIssuer` are also useful
- We might want to consider utility functions for the supported issuers/claims (e.g., `check_google_credential(id: &VerifiedID, id: String): bool`) to demonstrate correct usage/make it convenient

Reply:
Thanks for the feedback! I'm not sure what you suggest the `check_google_credential` function should do exactly. Could you elaborate?

Reply:
Ah, I was imagining something like:

/// The name of the key claim
        kc_name: String,

        /// The value of the key claim
        kc_value: String,

        /// The issuer
        iss: String,

        /// The audience (wallet)
        aud: String,
```
const EMAIL_KC_NAME = ...;

const GOOGLE_ISS = ...;

public fun check_google_email(id: &VerifiedID, email: String): bool {
  return id.iss == GOOGLE_ISS && id.kc_name == EMAIL_KC_NAME && id.kc_value = email
}
```

Basically, make it easier for application code to ask questions like "was this address derived from the Google email `email`?" without having to understand JWT details like the issuer constant or `kc_value`.

Reply:
Ah, that definitely makes sense to do. Great idea. In the case with google, the key claim is a user name and not an email in the test vectors I've found, but I'm not sure that's always the case. I think we should look into what is possible for each issuer adding these helper functions, so perhaps it makes sense to do that in a later PR?